### PR TITLE
`AddCommand` parse names as name capitalized by default and raise specific error for missing `Prefix`, make `Prefix` tags case-insensitive, stack `ParserException` messages

### LINKED ISSUE #134: Add command `Name` duplicate detection fix
`david Li` and `David Li` are considered different, consider manipulating the case for user when parsing.

### LINKED ISSUE #140: Add command - Handle missing prefix, tell user which missing prefix
For example, the add function missing income field 
`add n/John Doe p/98765432 e/johnd@example.com a/311, Clementi Ave 2, #02-25 f/4 t/buyer t/seller r/Has 3 cats.`
should give more specific error message than

"invalid command format! 
add: Adds a client to the Realodex. Parameters: n/NAME p/PHONE i/INCOME e/EMAIL a/ADDRESS f/FAMILY t/BUYER/SELLER [r/REMARK]"

### LINKED ISSUE #141: Add command - Make tags not case sensitive

### LINKED ISSUE #152: `add` should stack parser errors
If there are multiple offending fields, we should return error for each of these fields and not just one of them.

### PR DESCRIPTION
**Pull request by UdhayaShan1**
# Enhanced Add Command
This pull request introduces enhancements to the AddCommand feature to handle cases where required prefixes are missing. It includes the following changes:

- Parsing of names with proper capitalization for users.
   - Addition of ParserUtil tests for parsing different capitalization cases.
   - Addition of a test to ensure AddCommandParser creates AddCommand properly.
   - Add and fix existing tests
- Implementation of basic exception throwing for missing prefixes in AddCommand.

  - Addition of a method to return a list of missing prefixes.
  - Improvement of error message with more clarity to show what is missing
  - Fix tests

- Make `Prefix` tags case-insensitive as well

- Stack `ParserException` messages as multiple fields could be input wrongly by user.

image

***

### CODE REVIEW

File: src/main/java/seedu/realodex/logic/Messages.java [hunk 1]
```diff
@@ -32,6 +34,10 @@ public static String getErrorMessageForDuplicatePrefixes(Prefix... duplicatePref
         return MESSAGE_DUPLICATE_FIELDS + String.join(" ", duplicateFields);
     }
 
+    public static String getErrorMessageForMissingPrefixes(String exceptionMessageForMissingPrefixes) {
+        return MESSAGE_MISSING_PREFIXES + exceptionMessageForMissingPrefixes;
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/Messages.java:38]: 👍

File: src/main/java/seedu/realodex/logic/parser/Prefix.java [hunk 2]
```diff
@@ -15,6 +20,24 @@ public String getPrefix() {
...
+     * @param argMultimap The argument multimap containing prefixes mapped to their respective values.
+     * @param listOfCompulsoryPrefix An array of Prefix objects representing compulsory prefixes.
+     * @return A string message listing the missing prefixes separated by commas.
+     */
+    public static String returnMessageOfMissingPrefixes(Map<Prefix, List<String>> argMultimap,
+                                                        Prefix[] listOfCompulsoryPrefix) {
+        List<String> missingPrefixes = Stream.of(listOfCompulsoryPrefix)
+                .filter(prefix -> !argMultimap.containsKey(prefix))
+                .map(Prefix::toString)
+                .collect(Collectors.toList());
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/parser/Prefix.java:37]: neat way to filter out the present prefixes!

File: src/main/java/seedu/realodex/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -50,9 +50,34 @@ public static Name parseName(String name) throws ParseException {
...
+        return new Name(capitalizeWords(trimmedName));
     }
 
+    /**
+     * Capitalizes the first letter of each word in the given sentence.
+     *
+     * @param sentence the input sentence
+     * @return the sentence with the first letter of each word capitalized
+     */
+    public static String capitalizeWords(String sentence) {
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/parser/ParserUtil.java:62]: 👍

File: src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java [hunk 4]
```diff
@@ -28,8 +28,10 @@
  */
 public class CommandTestUtil {
 
-    public static final String VALID_NAME_AMY = "Amy Bee";
-    public static final String VALID_NAME_BOB = "Bob Choo";
+    public static final String VALID_NAME_AMY_CAPS = "Amy Bee";
+
+    public static final String VALID_NAME_AMY_NON_CAPS = "amy bee";
+    public static final String VALID_NAME_BOB_CAPS = "Bob Choo";
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:34]: Suggestion: Would it be ok to include tests also for all caps/irregular cased names? "AMY CHOO"/"aMY chOO" should both be parsed to "Amy Choo". To do this, perhaps you can consider changing the capitalizeWord method to toLower() everything first before capitalising the first word 😄

Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:34]: Good catch, my caps function was rather wrong too

Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:35]: Changed function and added new commits

File: src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java [hunk 5]
```diff
@@ -28,8 +28,11 @@
  */
 public class CommandTestUtil {
 
-    public static final String VALID_NAME_AMY = "Amy Bee";
-    public static final String VALID_NAME_BOB = "Bob Choo";
+    public static final String VALID_NAME_AMY_CAPS = "Amy Bee";
```
Comment (4llysa) [inline src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:31]: "CAPS" seems so suggest that the name should be in caps but its not? maybe replace with "AMY BEE"
```suggestion
    public static final String VALID_NAME_AMY_CAPS = "AMY BEE";
```

File: src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java [hunk 6]
```diff
@@ -28,8 +28,11 @@
...
+
+    public static final String VALID_NAME_AMY_NON_CAPS = "amy bee";
+    public static final String VALID_NAME_AMY_VARYING_CAPS = "aMY bEe";
+    public static final String VALID_NAME_BOB_CAPS = "Bob Choo";
```
Comment (4llysa) [inline src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:35]: Same here 
```suggestion
    public static final String VALID_NAME_BOB_CAPS = "BOB CHOO";
```

File: src/test/java/seedu/realodex/logic/parser/ParserUtilTest.java [hunk 7]
```diff
@@ -73,29 +77,78 @@ public void parseIndex_validInput_success() throws Exception {
...
+        assertEquals(expectedName, ParserUtil.parseName(nonCapitalizedInBothNames));
+    }
+
+    @Test
+    public void parseName_nonCapitalizedInFirstPartOfName_returnsCapitalizedName() throws Exception {
+        String nonCapitalizedInFirstNames = VALID_NAME_NON_CAPS_FIRST_NAME;
+        Name expectedName = new Name(ParserUtil.capitalizeWords(nonCapitalizedInFirstNames));
+        assertEquals(expectedName, ParserUtil.parseName(nonCapitalizedInFirstNames));
+    }
+
```
Comment (4llysa) [inline src/test/java/seedu/realodex/logic/parser/ParserUtilTest.java:151]: good job on the very extensive testing for parsing names

File: src/main/java/seedu/realodex/logic/parser/CliSyntax.java [hunk 8]
```diff
@@ -6,18 +6,18 @@
...
-    public static final Prefix PREFIX_NAME = new Prefix("n/");
-    public static final Prefix PREFIX_PHONE = new Prefix("p/");
-    public static final Prefix PREFIX_INCOME = new Prefix("i/");
-    public static final Prefix PREFIX_ADDRESS = new Prefix("a/");
+    public static final Prefix PREFIX_NAME = new Prefix("n/", "name");
+    public static final Prefix PREFIX_PHONE = new Prefix("p/", "phone");
+    public static final Prefix PREFIX_INCOME = new Prefix("i/", "income");
+    public static final Prefix PREFIX_ADDRESS = new Prefix("a/", "address");
 
     //To remove EMAIl field soon
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/CliSyntax.java:14]: are we removing email field soon?

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java [hunk 9]
```diff
@@ -40,40 +40,101 @@ public AddCommand parse(String args) throws ParseException {
...
+
+        Remark remark;
+        remark = ParserUtil.parseRemark(argMultimap.getRemarkValue(PREFIX_REMARK).orElseThrow());
+
+        // If any parsing operation fails, throw a ParseException with the accumulated error messages
+        if (errorMessageBuilder.length() > 0) {
+            throw new ParseException(errorMessageBuilder.toString().trim());
+        }
+
+        // If all parsing operations succeed, create the person object
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:122]: this has become a very long method. is there a way to shorten this using streams?

File: src/main/java/seedu/realodex/logic/parser/ArgumentMultimap.java [hunk 10]
```diff
@@ -80,4 +80,26 @@ public void verifyNoDuplicatePrefixesFor(Prefix... prefixes) throws ParseExcepti
...
+        return Prefix.returnMessageOfMissingPrefixes(argMultimap, listOfCompulsoryPrefix);
+    }
+
+    /**
+     * Returns an array of unique Prefix objects from the provided list of prefixes.
+     *
+     * @param prefixes Variable number of Prefix objects to be filtered for uniqueness.
+     * @return An array of unique Prefix objects.
+     */
+    public Prefix[] returnListOfCompulsoryTags(Prefix... prefixes) {
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/ArgumentMultimap.java:100]: good use of streams! 👍🏻

File: src/test/java/seedu/realodex/logic/parser/ArgumentTokenizerTest.java [hunk 11]
```diff
@@ -138,13 +138,13 @@ public void tokenize_multipleArgumentsJoined() {
...
-        Prefix aaa = new Prefix("aaa");
+        Prefix aaa = new Prefix("aaa", "aaa");
 
         assertEquals(aaa, aaa);
-        assertEquals(aaa, new Prefix("aaa"));
+        assertEquals(aaa, new Prefix("aaa", "aaa"));
 
         assertNotEquals(aaa, "aaa");
-        assertNotEquals(aaa, new Prefix("aab"));
+        assertNotEquals(aaa, new Prefix("aab", "aab"));
```
Comment (4llysa) [inline src/test/java/seedu/realodex/logic/parser/ArgumentTokenizerTest.java:147]: i might have missed it but i don't see any tests for checking that prefix is now case insensitive?

File: src/main/java/seedu/realodex/logic/parser/Prefix.java [hunk 12]
```diff
@@ -1,30 +1,91 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
 /**
  * A prefix that marks the beginning of an argument in an arguments string.
  * E.g. 't/' in 'add James t/ friend'.
  */
 public class Prefix {
     private final String prefix;
+    private final String prefixRepresentation;
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/Prefix.java:14]: perhaps change "prefixRepresentation" to something more intuitive? like prefixDescription? idk

File: src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java [hunk 13]
```diff
@@ -48,10 +48,19 @@ private static List<PrefixPosition> findPrefixPositions(String argsString, Prefi
         List<PrefixPosition> positions = new ArrayList<>();
 
         int prefixPosition = findPrefixPosition(argsString, prefix.getPrefix(), 0);
+        int prefixPositionOfUpper = findPrefixPosition(argsString, prefix.getPrefix().toUpperCase(), 0);
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java:51]: good job! remember to test this

File: src/main/java/seedu/realodex/logic/parser/CliSyntax.java (hunk 8 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/CliSyntax.java:14]: Nope, stray comment forgot remove

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java (hunk 9 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:122]: Not too sure GPT not helpful for this

File: src/test/java/seedu/realodex/logic/parser/ArgumentTokenizerTest.java (hunk 11 above)
Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/logic/parser/ArgumentTokenizerTest.java:147]: Nope no tests yet

File: src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java (hunk 13 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java:51]: Thanks!

File: src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java (hunk 5 above)
Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/logic/commands/CommandTestUtil.java:31]: I will change to `VALID_NAME_AMY_FIRST_LETTER_CAPS`

File: src/main/java/seedu/realodex/logic/parser/CliSyntax.java (hunk 8 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/CliSyntax.java:14]: Removed thanks!

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java (hunk 9 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:122]: I will think abt it

File: src/main/java/seedu/realodex/logic/parser/ArgumentMultimap.java (hunk 10 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/ArgumentMultimap.java:100]: Thanks!

File: src/main/java/seedu/realodex/logic/parser/Prefix.java (hunk 12 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/Prefix.java:14]: Changed thanks!

**Comment by UdhayaShan1** [comment]
Dont merge yet, I will try shorten `AddCommandParser` and add tests for case-insensitive prefixes.

**Comment by UdhayaShan1** [comment]
Shortened a bit, still need add tests

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java (hunk 9 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:106]: used generic class to encapsulate

File: src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java (hunk 13 above)
Comment (jyztintan) [inline src/main/java/seedu/realodex/logic/parser/ArgumentTokenizer.java:51]: nice, yep can just add more test cases in future as warned by codecov

File: src/main/java/seedu/realodex/logic/parser/ParserUtil.java [hunk 14]
```diff
@@ -50,9 +51,51 @@ public static Name parseName(String name) throws ParseException {
...
+
+    /**
+     * Parses a {@code String name} into a {@code Name} and returns it along with any parsing exception message.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @param name The name string to parse.
+     * @return A ParserResult containing the parsed Name or an exception message.
+     */
+    public static ParserResult<Name> parseNameReturnStored(String name) throws ParseException {
+        requireNonNull(name);
```
Comment (jyztintan) [inline src/main/java/seedu/realodex/logic/parser/ParserUtil.java:65]: good check!

File: src/main/java/seedu/realodex/logic/parser/Prefix.java [hunk 15]
```diff
@@ -1,30 +1,91 @@
...
  * A prefix that marks the beginning of an argument in an arguments string.
  * E.g. 't/' in 'add James t/ friend'.
  */
 public class Prefix {
     private final String prefix;
+    private final String prefixDescription;
 
-    public Prefix(String prefix) {
+    /**
+     * Constructs a Prefix object with the given prefix and its representation.
```
Comment (jyztintan) [inline src/main/java/seedu/realodex/logic/parser/Prefix.java:17]: Thanks for adding javadocs too. Succinct and clear 👍

File: src/main/java/seedu/realodex/logic/parser/Prefix.java [hunk 16]
```diff
@@ -1,30 +1,91 @@
...
+     *
+     * @param prefix The prefix string.
+     * @param prefixDescription The string representation of the prefix.
+     */
+    public Prefix(String prefix, String prefixDescription) {
         this.prefix = prefix;
+        this.prefixDescription = prefixDescription.toUpperCase();
```
Comment (jyztintan) [inline src/main/java/seedu/realodex/logic/parser/Prefix.java:24]: Good catch to do toUpperCase to prevent bugs caused by case-sensitivity

**Review by jyztintan — APPROVED** [review]
LGTM!

**Comment by UdhayaShan1** [comment]
Ok done, all tests added

**Pull request merged**